src/tcp_sock.rs

```diff
@@ -251,59 +238,24 @@ impl Inner {
     //                 will be attempted in the next write schedule.
     //   - Err(error): there was an error while writing to the socket.
     fn write<T: Serialize>(&mut self, msg: Option<(T, Priority)>) -> ::Res<bool> {
-        let expired_keys: Vec<u8> = self
-            .write_queue
-            .iter()
-            .skip_while(|&(&priority, queue)| {
-                priority < MSG_DROP_PRIORITY || // Don't drop high-priority messages.
-                queue.front().map_or(true, |&(ref timestamp, _)| {
-                    timestamp.elapsed().as_secs() <= MAX_MSG_AGE_SECS
-                })
-            }).map(|(&priority, _)| priority)
-            .collect();
-        let dropped_msgs: usize = expired_keys
-            .iter()
-            .filter_map(|priority| self.write_queue.remove(priority))
-            .map(|queue| queue.len())
-            .sum();
-        if dropped_msgs > 0 {
-            trace!(
-                "Insufficient bandwidth. Dropping {} messages with priority >= {}.",
-                dropped_msgs,
-                expired_keys[0]
-            );
-        }
-
+        let _ = self.out_queue.drop_expired();
         if let Some((msg, priority)) = msg {
-            let mut data = Cursor::new(Vec::with_capacity(mem::size_of::<u32>()));
-
-            let _ = data.write_u32::<LittleEndian>(0);
-
-            serialise_into(&msg, &mut data)?;
-
-            let len = data.position() - mem::size_of::<u32>() as u64;
-            data.set_position(0);
-            data.write_u32::<LittleEndian>(len as u32)?;
-
-            let entry = self
-                .write_queue
-                .entry(priority)
-                .or_insert_with(|| VecDeque::with_capacity(10));
-            entry.push_back((Instant::now(), data.into_inner()));
+            self.enqueue_data(msg, priority)?;
         }
+        self.flush_write_until_would_block()
+    }
 
+    /// Returns `Ok(false)`, if write to the underlying stream would block.
+    fn flush_write_until_would_block(&mut self) -> ::Res<bool> {
```

Comment:
I definitely don't want to be very precise, I want functions to leak as less implementation details as possible. Therefore saying `until_would_block` sounds better to me as it is more like a natural language rather than implementation specific.
`man 2 recv` says smth like along the lines:
```
in which case the
       value  -1  is  returned  and  the  external  variable  errno is set to EAGAIN or EWOULDBLOCK.
```
So the error message could be other than `EWOULDBLOCK`.